File: team2-backend/src/models/audit_log.py
"""AuditLog model — DATA-04 §4 audit_logs."""
from datetime import datetime, timezone
from typing import Optional

from sqlmodel import Field, SQLModel


def utcnow() -> str:
    return datetime.now(timezone.utc).isoformat()


class AuditLog(SQLModel, table=True):
    __tablename__ = "audit_logs"

    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.user_id")
    entity_type: str = Field(nullable=False, max_length=64)
    entity_id: Optional[int] = None
    action: str = Field(nullable=False, max_length=64)
    detail: Optional[str] = None  # TEXT (JSON)
    ip_address: Optional[str] = None
    created_at: str = Field(default_factory=utcnow)

File: team1-frontend/lib/models/entities/audit_log.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'audit_log.freezed.dart';
part 'audit_log.g.dart';

@freezed
class AuditLog with _$AuditLog {
  const factory AuditLog({
    required int id,
    @JsonKey(name: 'userId') required int userId,
    @JsonKey(name: 'entityType') required String entityType,
    @JsonKey(name: 'entityId') int? entityId,
    required String action,
    String? detail,
    @JsonKey(name: 'ipAddress') String? ipAddress,
    @JsonKey(name: 'createdAt') required String createdAt,
  }) = _AuditLog;

  factory AuditLog.fromJson(Map<String, dynamic> json) =>
      _$AuditLogFromJson(json);
}
